datadog/resource_datadog_dashboard_json.go

```diff
@@ -0,0 +1,168 @@
+package datadog
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/structure"
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"
+)
+
+var computedFields = []string{"id", "author_handle", "author_name", "created_at", "modified_at", "url"}
+
+func resourceDatadogDashboardJSON() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDatadogDashboardJSONCreate,
+		Read:   resourceDatadogDashboardJSONRead,
+		Update: resourceDatadogDashboardJSONUpdate,
+		Delete: resourceDatadogDashboardJSONDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"dashboard": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (warns []string, errs []error) {
+					_, err := utils.NormalizeJSONorYamlString(v.(string))
+					if err != nil {
+						errs = append(errs, fmt.Errorf("invalid JSON or YAML: \n %s", v))
+					}
+					return
+				},
+				StateFunc: func(v interface{}) string {
+					k, _ := utils.NormalizeJSONorYamlString(v.(string))
+					// Remove computed fields when comparing diffs
+					attrMap, _ := structure.ExpandJsonFromString(k)
+					for _, f := range computedFields {
+						delete(attrMap, f)
+					}
+					res, _ := structure.FlattenJsonToString(attrMap)
+					return res
+				},
+				Description: "The JSON or YAML formatted definition of the Dashboard.",
+			},
+			"url": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "The URL of the dashboard.",
+			},
+		},
+	}
+}
+
+func resourceDatadogDashboardJSONRead(d *schema.ResourceData, meta interface{}) error {
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV1 := providerConf.DatadogClientV1
+	authV1 := providerConf.AuthV1
+	path := "/api/v1/dashboard/" + d.Id()
+
+	respByte, httpresp, err := utils.SendRequest(authV1, datadogClientV1, "GET", path, nil)
+	if err != nil {
+		if httpresp != nil && httpresp.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	respMap, err := utils.ConvertResponseByteToMap(respByte)
+	if err != nil {
+		return err
+	}
+
+	return updateDashboardJSONState(d, respMap)
+}
+
+func resourceDatadogDashboardJSONCreate(d *schema.ResourceData, meta interface{}) error {
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV1 := providerConf.DatadogClientV1
+	authV1 := providerConf.AuthV1
+	path := "/api/v1/dashboard"
+
+	test, err := utils.NormalizeJSONorYamlString(d.Get("dashboard").(string))
```

Comment:
That's an interesting approach. FWIW my thinking was more to check if calling `jsonencode(yamldecode(data)` would work. Do you think that would simplify the code?

Reply:
Also let's name the variable something else than `test`.